diff --git a/include/sys/mman.h b/include/sys/mman.h
new file mode 100644
index 0000000..fc78533
--- /dev/null
+++ b/include/sys/mman.h
@@ -0,0 +1,35 @@
+///////////////////////////////////////////////////////////////////////////////
+// Licensed Materials - Property of IBM
+// ZOSLIB
+// (C) Copyright IBM Corp. 2020. All Rights Reserved.
+// US Government Users Restricted Rights - Use, duplication
+// or disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
+///////////////////////////////////////////////////////////////////////////////
+
+#include_next <sys/mman.h>
+#ifdef __cplusplus
+  extern "C" {
+#endif
+
+void *zoslib_mmap(void* addr, size_t length, int prot, int flags, int fd, off_t offset);
+int zoslib_munmap(void *addr, size_t len);
+
+#pragma map(zos_mmap, "mmap")
+#pragma map(zos_munmap, "munmap")
+void *zos_mmap(void* addr, size_t len, int prot, int flags, int fd, off_t offset);
+int zos_munmap(void *addr, size_t len);
+
+#ifdef __cplusplus
+  }
+#endif
+
+#define mmap (zoslib_mmap)
+#define munmap (zoslib_munmap)
+
+#ifndef MAP_ANONYMOUS
+  #define MAP_ANONYMOUS (16)
+#endif
+
+#ifndef MAP_ANON
+  #define MAP_ANON (MAP_ANONYMOUS)
+#endif
