diff --git a/test/test-mmap.cc b/test/test-mmap.cc
new file mode 100644
index 0000000..7baac8f
--- /dev/null
+++ b/test/test-mmap.cc
@@ -0,0 +1,29 @@
+#include "zos.h"
+#include "gtest/gtest.h"
+#include <libgen.h>
+#include <sys/mman.h>
+
+namespace {
+
+TEST(NewFunctionsTest, AnonymousMMapAndMUNMapLeakDetection) {
+  //
+  // Ensure I can map / unmap quite a few times without exhausting storage
+  // (i.e. no memory leak) 
+  void* memory;
+  char* cmem;
+
+  for (int i=0; i<100; ++i) {
+    // allocate 1GB and then update a byte on the start and end, 
+    // then delete and repeat
+    memory = mmap(0, 1<<19, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0);
+    cmem = (char*) memory;
+    EXPECT_STRNE(cmem, NULL);
+
+    cmem[0] = cmem[1<<19 -1] = 'X';
+
+    munmap(cmem, 1<<19);
+  }
+
+}
+
+} // namespace
